ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/RemoteConnectorExecutor.java

```diff
@@ -188,16 +199,65 @@ && getUserRateLimiterMap().get(user.getName()) != null
             if (getMlGuard() != null && !getMlGuard().validate(payload, MLGuard.Type.INPUT)) {
                 throw new IllegalArgumentException("guardrails triggered for user input");
             }
-            invokeRemoteModel(mlInput, parameters, payload, tensorOutputs, countDownLatch, actionListener);
+            if (executionContext.getConnectorRetryOption().isRetryEnabled()) {
+                invokeRemoteModelWithRetry(mlInput, parameters, payload, executionContext, actionListener);
+            } else {
+                invokeRemoteModel(mlInput, parameters, payload, executionContext, actionListener);
+            }
         }
     }
 
+    default void invokeRemoteModelWithRetry(
+        MLInput mlInput,
+        Map<String, String> parameters,
+        String payload,
+        ExecutionContext executionContext,
+        ActionListener<Tuple<Integer, ModelTensors>> actionListener
+    ) {
+        final RetryableAction<Tuple<Integer, ModelTensors>> invokeRemoteModelAction = new RetryableAction<>(
+            getLogger(),
+            getClient().threadPool(),
+            TimeValue.timeValueMillis(executionContext.getConnectorRetryOption().getRetryBackoffMillis()),
+            TimeValue.timeValueSeconds(executionContext.getConnectorRetryOption().getRetryTimeoutSeconds()),
+            actionListener,
+            BackoffPolicy
+                .constantBackoff(
+                    TimeValue.timeValueMillis(executionContext.getConnectorRetryOption().getRetryBackoffMillis()),
+                    Integer.MAX_VALUE
+                ),
+            executionContext.getConnectorRetryOption().getRetryExecutor()
+        ) {
+            Integer retryTimes = 0;
+
+            @Override
+            public void tryAction(ActionListener<Tuple<Integer, ModelTensors>> listener) {
+                // the listener here is RetryingListener
+                // If the request success, or can not retry, will call delegate listener
+                invokeRemoteModel(mlInput, parameters, payload, executionContext, listener);
+            }
+
+            @Override
+            public boolean shouldRetry(Exception e) {
+                Throwable cause = ExceptionsHelper.unwrapCause(e);
+                Integer maxRetryTimes = executionContext.getConnectorRetryOption().getMaxRetryTimes();
+                boolean shouldRetry = cause instanceof RetryableException;
+                if (shouldRetry) {
+                    getLogger().debug(String.format(Locale.ROOT, "The %d-th retry for invoke remote model", ++retryTimes));
```

Comment:
Logging here is misleading, consider the first time: the log says `The 1th retry for invoke remote model`, but if the maxRetryTimes is -1, the retry doesn't happen.

Reply:
Fixed in the new commit.
BTW if the maxRetryTimes is -1, we don't restrict it, we only restrict by time out value. But the log here does mislead users and I just fixed it